statsmodels/base/_penalties.py

```diff
@@ -110,13 +110,190 @@ def func(self, params):
+        tau = self.tau
+        p = np.atleast_1d(params)
+        p_abs = np.abs(p)
+        p_sign = np.sign(p)
+        res = np.zeros(p_abs.shape)#, p_abs.dtype)
+
+        mask1 = p_abs < tau
+        mask3 = p_abs >= self.c * tau
+        mask2 = ~mask1 & ~mask3
+
+        res[mask2] = -p_sign[mask2] / (self.c - 1)
+
+        return self.weights * res
+
+
+class SCADSmoothed(SCAD):
+    """
+    The SCAD penalty of Fan and Li, quadratically smoothed around zero
+
+    This follows Fan and Li 2001 equation (3.7).
+
+    Parameterization follows Boo, Johnson, Li and Tan 2011
+
+
+    TODO: Use delegation instead of subclassing, so smoothing can be added to all
+    penalty classes.
+
+    """
+
+    def __init__(self, tau, c=3.7, c0=None, wts=None, restriction=None):
+        if wts is None:
+            self.weights = 1.
+        else:
+            self.weights = wts
+        self.alpha = 1.
+        self.tau = tau
+        self.c = c
+        self.c0 = c0 if c0 is not None else tau * 0.1
+        if self.c0 > tau:
+            raise ValueError('c0 cannot be larger than tau')
+
+        # get coefficients for quadratic approximation
+        c0 = self.c0
+        deriv_c0 = super(SCADSmoothed, self).grad(c0)
+        value_c0 = super(SCADSmoothed, self).func(c0)
+        self.aq1 = value_c0 - 0.5 * deriv_c0 * c0
+        self.aq2 = 0.5 * deriv_c0 / c0
+        self.restriction = restriction
+
+
+    def func(self, params):
+        # TODO: `and np.size(params) > 1` is hack for llnull, need better solution
+        if self.restriction is not None and np.size(params) > 1:
+            params = self.restriction.dot(params)
+        # need to temporarily override weights for call to super
+        # Note: we have the same problem with `restriction`
+        weights = self.weights
+        self.weights = 1.
+        value = super(SCADSmoothed, self).func(params[None, ...])
+        self.weights = weights
+
+        #change the segment corrsponding to quadratic approximation
+        p_abs = np.atleast_1d(np.abs(params))
+        mask = p_abs < self.c0
+        p_abs_masked = p_abs[mask]
+        value[mask] = self.aq1 + self.aq2 * p_abs_masked**2
+
+        return (self.weights * value).sum(0)
+
+
+    def grad(self, params):
+        if self.restriction is not None and np.size(params) > 1:
+            params = self.restriction.dot(params)
+        # need to temporarily override weights for call to super
+        weights = self.weights
+        self.weights = 1.
+        value = super(SCADSmoothed, self).grad(params)
+        self.weights = weights
+
+        #change the segment corrsponding to quadratic approximation
+        p = np.atleast_1d(params)
+        mask = np.abs(p) < self.c0
+        value[mask] = 2 * self.aq2 * p[mask]
+
+        if self.restriction is not None and np.size(params) > 1:
+            return weights * value.dot(self.restriction)
+        else:
+            return weights * value
+
```

Comment:
extra whitespace

Reply:
I don't see any extra whitespace

Reply:
Must have been fixed in an intermediate commit.  Regardless, thanks for addressing the various commented-out code requests.